db/change_cache.go

```diff
@@ -577,31 +577,118 @@ func (c *changeCache) releaseUnusedSequence(ctx context.Context, sequence uint64
 	}
 }
 
-// releaseUnusedSequenceRange calls processEntry for each sequence in the range, but only issues a single notify.
+// releaseUnusedSequenceRange will handle unused sequence range arriving over DCP. It will batch remove from skipped or
+// push a range to pending sequences, or both.
 func (c *changeCache) releaseUnusedSequenceRange(ctx context.Context, fromSequence uint64, toSequence uint64, timeReceived time.Time) {
 
 	base.InfofCtx(ctx, base.KeyCache, "Received #%d-#%d (unused sequence range)", fromSequence, toSequence)
 
 	unusedSeq := channels.NewID(unusedSeqKey, unusedSeqCollectionID)
 	allChangedChannels := channels.SetOfNoValidate(unusedSeq)
-	for sequence := fromSequence; sequence <= toSequence; sequence++ {
+
+	// if range is single value, just run sequence through process entry and return early
+	if fromSequence == toSequence {
 		change := &LogEntry{
-			Sequence:     sequence,
+			Sequence:     toSequence,
 			TimeReceived: timeReceived,
 		}
-
-		// Since processEntry may unblock pending sequences, if there were any changed channels we need
-		// to notify any change listeners that are working changes feeds for these channels
 		changedChannels := c.processEntry(ctx, change)
 		allChangedChannels = allChangedChannels.Update(changedChannels)
 		c.channelCache.AddUnusedSequence(change)
+		if c.notifyChange != nil {
+			c.notifyChange(ctx, allChangedChannels)
+		}
+		return
 	}
 
+	// acquire mutex on cache for reading/writing to pending + reading next sequence value
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if toSequence < c.nextSequence {
+		// batch remove from skipped
+		err := c.RemoveSkippedSequences(ctx, fromSequence, toSequence)
+		if err != nil {
+			base.DebugfCtx(ctx, base.KeyCache, err.Error())
+			// we have sequences (possibly duplicate sequences) in the unused range that don't exist in skipped list
+			seqRange := &channels.UnusedSequenceRange{
+				StartSeq: fromSequence,
+				EndSeq:   toSequence,
+			}
+			// handle any potential duplicate sequences between the unused range and the skipped list
+			err = c.skippedSeqs._removeSubsetOfRangeFromSkipped(ctx, seqRange)
+			if err != nil {
+				base.DebugfCtx(ctx, base.KeyCache, "error processing subset of unused sequences in skipped list: %v", err)
+			}
+		}
+	} else if fromSequence >= c.nextSequence {
+		// whole range to pending
+		seqRange := &channels.UnusedSequenceRange{
+			StartSeq: fromSequence,
+			EndSeq:   toSequence,
+		}
+		c._pushRangeToPending(ctx, seqRange, timeReceived)
+		// unblock any pending sequences we can after new range(s) have been pushed to pending
+		changedChannels := c._addPendingLogs(ctx)
+		allChangedChannels = allChangedChannels.Update(changedChannels)
+		c.internalStats.pendingSeqLen = len(c.pendingLogs)
+	} else {
+		base.DebugfCtx(ctx, base.KeyCache, "unused sequence range of #%d to %d contains duplicate sequences", fromSequence, toSequence)
```

Comment:
Is there a reason we shouldn't split the unused sequence range at c.nextSequence, and send the first part through the skipped handling and the second part through the pending handling?

If we can't do that, I think the log message should be higher (warn or info), and should indicate that this sequence range won't be removed from skipped until they are abandoned.

Reply:
Its definitely possible to do so, my code before was doing that but removed it as I thought we cam to decision it shouldn't be possible for that scenario to present itself. Should we handle it anyway? 

I have raised the logging to warn for now but happy to make changes to split the ranges if you think its worth doing still?